src/connection.rs

```diff
@@ -29,85 +29,53 @@ pub enum ForwardType {
-    {
-        self.transformer.get_buffer().clear();
-        read_exact_to_vec(
-            &mut self.raw_conn.stream,
-            self.transformer.get_buffer(),
-            size,
-        )
-        .await?;
+    pub(crate) async fn read_response(&mut self) -> Result<Response> {
+        let buffer = &mut self.read_buffer;
 
-        self.deserialize()
-    }
+        if buffer.len() < 4 {
+            let n = 4 - buffer.len();
+            read_to_vec_rng(&mut self.raw_conn.stream, buffer, n..).await?;
+        }
 
-    /// Return size of the response.
-    async fn read_header(&mut self) -> Result<u32> {
-        self.read_and_deserialize(4).await
-    }
+        // Read in the header
+        let packet_len: u32 = Self::deserialize(&buffer[..4])?;
 
-    pub(crate) async fn read_response(&mut self) -> Result<Response> {
-        let len = self.read_header().await?;
-        self.read_and_deserialize(len as usize).await
-    }
+        let packet_len: usize = packet_len.try_into().unwrap();
 
-    fn try_read_and_deserialize<'a, T>(&'a mut self, size: usize) -> Result<Option<T>>
-    where
-        T: Deserialize<'a>,
-    {
-        self.transformer.get_buffer().resize(size, 0);
+        // The first 4 bytes are not counted as the packet body
+        let buffer_len = buffer.len() - 4;
 
-        // Ignore any trailing bytes to be forward compatible
-        match self.raw_conn.try_read(self.transformer.get_buffer())? {
-            Some(_) => Ok(Some(self.deserialize()?)),
-            None => Ok(None),
+        if buffer_len < packet_len {
+            // Read in rest of the packet
+            let n = packet_len - buffer_len;
+            read_to_vec_rng(&mut self.raw_conn.stream, buffer, n..).await?;
         }
-    }
 
-    fn try_read_header(&mut self) -> Result<Option<u32>> {
-        self.try_read_and_deserialize(4)
-    }
+        // Deserialize the response
+        let response = Self::deserialize(&buffer[4..(4 + packet_len)])?;
 
-    /// If it is readable, then the entire packet can be read in blocking manner.
-    ///
-    /// While this is indeed a blocking call, it is unlikely to block since ssh mux
-    /// master most likely would send it using one write/send.
-    ///
-    /// Even if it does employ multiple write/send, these functions would just return
-    /// immediately since the buffer for the unix socket is empty and should be big
-    /// enough for one message.
-    ///
-    /// If it is not readable, then it would return Ok(None).
-    pub(crate) fn try_read_response(&mut self) -> Result<Option<Response>> {
-        let len = match self.try_read_header()? {
-            Some(len) => len as usize,
-            None => return Ok(None),
-        };
-        loop {
-            if let Some(response) = self.try_read_and_deserialize(len)? {
-                break Ok(Some(response));
-            }
-        }
+        // Remove the packet from buffer
+        buffer.drain(..(4 + packet_len));
```

Comment:
This is still potentially going to be pretty expensive, since it requires shifting around all the remaining bytes of the `Vec`. You could try using a `VecDeque` instead, which doesn't have this problem, although then you have to deal with wrap-around, which may also be fairly painful.

Reply:
That should be OK because most of the time, we receive one response packet after sending one request packet.

Only for a new session opened, can the mux server returns two packet in a row, but most of the time, it would take a long time before the second packet come through because the ssh mux server needs to communicate with the destination.